### PR TITLE
[wasi] Add `wasi-experimental` workload, `wasiconsole` template, and `Wasi.Build.Tests`

### PR DESCRIPTION
**Pull request by radical**
- Initial `Wasi.Build.Tests`, with some duplicated code between Wasi, and Wasm build tests. This is being cleaned in a separate branch for a follow up PR
- Add a new `wasiconsole` template
- WasmAppHost - add support to run with `wasmtime`

### CODE REVIEW

File: eng/liveBuilds.targets [hunk 1]
```diff
@@ -222,6 +222,13 @@
     </ItemGroup>
 
     <ItemGroup Label="Wasi" Condition="'$(TargetOS)' == 'wasi'">
+      <LibrariesRuntimeFiles
+                             Include="
+        $(LibrariesNativeArtifactsPath)dotnet.wasm;
+        $(LibrariesNativeArtifactsPath)dotnet.timezones.blat;
```
Comment (pavelsavara) [inline eng/liveBuilds.targets:228]: we will clash here with my PR on dotnet.timezones.blat;

File: src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Current.Manifest/WorkloadManifest.targets.in [hunk 2]
```diff
@@ -111,13 +113,23 @@
...
+  </ImportGroup>
+
+  <PropertyGroup Condition="'$(TargetsCurrent)' == 'true' and ('$(TargetPlatformIdentifier)' == 'android' or '$(TargetPlatformIdentifier)' == 'macOS' or '$(TargetPlatformIdentifier)' == 'ios' or '$(TargetPlatformIdentifier)' == 'maccatalyst' or '$(TargetPlatformIdentifier)' == 'tvos' or '$(RuntimeIdentifier)' == 'browser-wasm' or '$(RuntimeIdentifier)' == 'wasi-wasm')">
     <_MonoWorkloadTargetsMobile>true</_MonoWorkloadTargetsMobile>
     <_MonoWorkloadRuntimePackPackageVersion>$(_RuntimePackInWorkloadVersionCurrent)</_MonoWorkloadRuntimePackPackageVersion>
   </PropertyGroup>
 
   <ItemGroup Condition="'$(TargetsCurrent)' == 'true' and '$(_MonoWorkloadTargetsMobile)' == 'true'">
     <KnownRuntimePack Update="@(KnownRuntimePack)">
+      <!-- FIXMEWASI: workaround for now -->
```
Comment (pavelsavara) [inline src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Current.Manifest/WorkloadManifest.targets.in:130]: TODOWASI for consistency

File: src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,713 @@
...
+                                                   string targetFramework,
+                                                   bool hasIcudt = true,
+                                                   bool dotnetWasmFromRuntimePack = true)
+        {
+#if false
+            AssertFilesExist(bundleDir, new []
+            {
+                "index.html",
+                mainJS,
+                "dotnet.timezones.blat",
```
Comment (pavelsavara) [inline src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs:363]: another clash

File: src/mono/wasm/templates/templates/wasi-console/wasi-console.0.csproj [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>netX.0</TargetFramework>
```
Comment (pavelsavara) [inline src/mono/wasm/templates/templates/wasi-console/wasi-console.0.csproj:3]: net8 ? should we update the browser templates to net8 too ?

File: src/tasks/WasmAppBuilder/wasi/EmitWasmBundleObjectFile.cs [hunk 5]
```diff
@@ -109,23 +110,55 @@ public override bool Execute()
...
             Arguments = $"-xc -o \"{destinationObjectFile}\" -c -",
             RedirectStandardInput = true,
+            RedirectStandardOutput = true,
+            RedirectStandardError = true,
             UseShellExecute = false,
         })!;
 
-        BundleFileToCSource(destinationObjectFile, fileToBundle, clangProcess.StandardInput.BaseStream);
-        clangProcess.WaitForExit();
+        clangProcess.ErrorDataReceived += (sender, e) =>
```
Comment (pavelsavara) [inline src/tasks/WasmAppBuilder/wasi/EmitWasmBundleObjectFile.cs:132]: My PR has different version of this.

File: src/mono/wasm/host/wasi/WasiEngineHost.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (pavelsavara) [inline src/mono/wasm/host/wasi/WasiEngineHost.cs:1]: Maybe wasi host should be separate project and separate host. 
There are dependencies which don't make sense for WASI like `BrowserDebugHost`

File: src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+                if (File.Exists(file))
+                    table.Add(Path.GetFileName(file), new FileStat(FullPath: file, Exists: true, LastWriteTimeUtc: File.GetLastWriteTimeUtc(file), Length: new FileInfo(file).Length));
+                else
+                    table.Add(Path.GetFileName(file), new FileStat(FullPath: file, Exists: false, LastWriteTimeUtc: DateTime.MinValue, Length: 0));
+            }
+
+            return table;
+        }
+
+        protected static string GetSkiaSharpReferenceItems()
```
Comment (pavelsavara) [inline src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs:670]: Does SkiaSharp run on wasi ?

File: src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+                                                   string mainJS,
+                                                   bool hasV8Script,
+                                                   string targetFramework,
+                                                   bool hasIcudt = true,
+                                                   bool dotnetWasmFromRuntimePack = true)
+        {
+#if false
+            AssertFilesExist(bundleDir, new []
+            {
+                "index.html",
```
Comment (pavelsavara) [inline src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs:361]: index.html should not be there on wasi

File: eng/liveBuilds.targets (hunk 1 above)
Comment (radical) [reply eng/liveBuilds.targets:228]: It won't be merged before the snap, correct? You can change it once this is merged as you see fit.

File: src/mono/wasm/host/wasi/WasiEngineHost.cs (hunk 6 above)
Comment (radical) [reply src/mono/wasm/host/wasi/WasiEngineHost.cs:1]: We can discuss that. For now I want to have a single WasmAppHost which can run any of the projects.

File: src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs (hunk 8 above)
Comment (radical) [reply src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs:361]: it's `#ifdef'ed` out

File: src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs (hunk 7 above)
Comment (radical) [reply src/mono/wasi/Wasi.Build.Tests/BuildTestBase.cs:670]: This, and other bits will get de-duplicated/cleaned up in a follow up PR.

File: src/mono/wasm/templates/templates/wasi-console/wasi-console.0.csproj (hunk 4 above)
Comment (radical) [reply src/mono/wasm/templates/templates/wasi-console/wasi-console.0.csproj:3]: `netX.0` is just the string that will get replaced by the actual tfm used.

**Review by lewing — APPROVED** [review]

**Comment by radical** [comment]
Failure is unrelated, and known.

**Pull request merged**